File: javascript/tests/after-refactor/DoubleArgumentMarshaler.test.js
const sinon = require('sinon');
const { ErrorCode } = require('../../after-refactor/ArgsError');
const { DoubleArgumentMarshaler } = require('../../after-refactor/DoubleArgumentMarshaler');

describe('#DoubleArgumentMarshaler', () => {
  describe('#set', () => {
    describe('Given a valid iterator', () => {
      it('Should read value from the iterator', () => {
        const mockIter = {
          next: sinon.stub(),
        };
        mockIter.next.returns({ done: false, value: '1.1' });
        const sut = new DoubleArgumentMarshaler();

        sut.set(mockIter);
        const value = DoubleArgumentMarshaler.getValue(sut);

        expect(mockIter.next.calledOnce).toBe(true);
        expect(value).toBe(1.1);
      });
    });

    describe('Given a iterator with bad float format', () => {
      it('Should throw an error', () => {
        const mockIter = {
          next: sinon.stub(),
        };
        mockIter.next.returns({ done: false, value: 'bad' });
        const sut = new DoubleArgumentMarshaler();

        let err;
        try {
          sut.set(mockIter);
        } catch (e) {
          err = e;
        }

        expect(mockIter.next.calledOnce).toBe(true);
        expect(err.getErrorCode()).toBe(ErrorCode.INVALID_DOUBLE);
      });
    });

    describe('Given a iterator with no more value', () => {
      it('Should throw an error', () => {
        const mockIter = {
          next: sinon.stub(),
        };
        mockIter.next.returns({ done: true });
        const sut = new DoubleArgumentMarshaler();

        let err;
        try {
          sut.set(mockIter);
        } catch (e) {
          err = e;
        }

        expect(mockIter.next.calledOnce).toBe(true);
        expect(err.getErrorCode()).toBe(ErrorCode.MISSING_DOUBLE);
      });
    });
  });

  describe('#getValue', () => {
    describe('Given a marshaler without calling set', () => {
      it('Should return the default value: 0', () => {
        const marshaler = new DoubleArgumentMarshaler();

        const value = DoubleArgumentMarshaler.getValue(marshaler);

        expect(value).toBe(0);
      });
    });
    describe('Given a non-marshaler object', () => {
      it('Should return the default value: 0', () => {
        const value = DoubleArgumentMarshaler.getValue(undefined);

        expect(value).toBe(0);
      });
    });
  });
});

File: java/src/test/java/args/after/DoubleArgumentMarshalerTest.java
package args.after;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.testng.Assert.assertEquals;
import static org.testng.Assert.fail;

import java.util.Iterator;
import java.util.NoSuchElementException;
import org.testng.annotations.Test;

/**
 * Unit test for DoubleArgumentMarshaler.
 */
public class DoubleArgumentMarshalerTest {
  @Test
  @SuppressWarnings(value = "unchecked")
  public void set_validInput() throws Exception {
    Iterator<String> mockIterator = (Iterator<String>) mock(Iterator.class);
    when(mockIterator.next()).thenReturn("1.1");
    when(mockIterator.hasNext()).thenReturn(true);
    DoubleArgumentMarshaler sut = new DoubleArgumentMarshaler();

    sut.set(mockIterator);
    double value = DoubleArgumentMarshaler.getValue(sut);

    assertEquals(value, 1.1, "Should return 1.1 after the marshaler reads from the iterator");
    verify(mockIterator, times(1).description("Should only read one element from the iterator"))
        .next();
  }

  @Test
  @SuppressWarnings(value = "unchecked")
  public void set_invalidInput() throws Exception {
    Iterator<String> mockIterator = (Iterator<String>) mock(Iterator.class);
    when(mockIterator.next()).thenReturn("bad");
    when(mockIterator.hasNext()).thenReturn(false);
    DoubleArgumentMarshaler sut = new DoubleArgumentMarshaler();

    testThrowArgsException(sut, mockIterator, ArgsErrorCode.INVALID_DOUBLE);
    verify(mockIterator, times(1).description("Should only read one element from the iterator"))
        .next();
  }

  private void testThrowArgsException(DoubleArgumentMarshaler marshaler, Iterator<String> input,
      ArgsErrorCode expectedCode) {
    try {
      marshaler.set(input);
    } catch (ArgsException e) {
      assertEquals(e.getErrorCode(), expectedCode);
      return;
    }
    fail(String.format("Expecting exception with code [%s], but got none", expectedCode.name()));
  }

  @Test
  @SuppressWarnings(value = "unchecked")
  public void set_noMoreInput() throws Exception {
    Iterator<String> mockIterator = (Iterator<String>) mock(Iterator.class);
    when(mockIterator.next()).thenThrow(new NoSuchElementException());
    when(mockIterator.hasNext()).thenReturn(true);
    DoubleArgumentMarshaler sut = new DoubleArgumentMarshaler();

    testThrowArgsException(sut, mockIterator, ArgsErrorCode.MISSING_DOUBLE);
    verify(mockIterator, times(1).description("Should only read one element from the iterator"))
        .next();
  }

  @Test
  public void getValue_beforeCallingSet() throws Exception {
    DoubleArgumentMarshaler sut = new DoubleArgumentMarshaler();

    double value = DoubleArgumentMarshaler.getValue(sut);

    assertEquals(value, 0, "Should return 0 before the marshaler takes any input");
  }

  @Test
  public void getValue_nullInput() throws Exception {
    double value = DoubleArgumentMarshaler.getValue(null);

    assertEquals(value, 0, "Should return 0 given no marshaler");
  }

  @Test
  public void getValue_invalidMarshaler() throws Exception {
    ArgumentMarshaler unknownMarshaler = new ArgumentMarshaler() {
      @Override
      public void set(Iterator<String> currentArgument) throws ArgsException {
        currentArgument.hasNext();
      }
    };
    double value = DoubleArgumentMarshaler.getValue(unknownMarshaler);

    assertEquals(value, 0, "Should return 0 given unknown marshaler");
  }
}
